Add salesforce authenticator.

weissleb opened the pull request:
This allows Presto admins to enable users to authenticate to Presto via Salesforce username and password/token. This is **not** a Salesforce connector, and does not allow users to query Salesforce data. It's simply a means by which users can authenticate to Presto, similar to LDAP or password file.

There is a single layer of security beyond simply providing username and password/token, which is that the admin can configure, via `salesforce.org` in `etc/password-authenticator.properties`, one or more 18 character Salesforce Organization Id's in a comma separated list. The user's successful login to Salesforce must return a matching Organization Id. Optionally, the admin can configure "all" to explicitly ignore this layer of security.

@lhofhansl

weissleb commented:
Looks like 4 checks failed, which don't appear related to my code.  Please advise if someone feels otherwise.

lhofhansl commented:
The tests look indeed unrelated, all Hive related.

weissleb commented:
@dain , @martint if you guys want to take this for a spin, you can sign up for a free Salesforce developer org.

1. Sign up at https://developer.salesforce.com/signup.
2. Login and reset your security token.
   * Upper right profile circle, choose settings.
   * Search Reset My Security Token.
   * Click Reset Security Token.
   * Check your email.
   * Your Presto user will be the Salesforce username, and your Presto password will be your Salesforce **password concatenated with your security token**.
3. Configure Presto with Org Id.
   * Upper right cog, choose Setup.
   * Search Company Information.
   * Locate the Salesforce.com Organization ID on the right-side column on the page.
   * Copy the ID and visit https://help.magentrix.com/aspx/ConvertSFID.  Paste the ID and convert it to the 18 char ID.
   * This 18 char ID goes into `salesforce.org` in `etc/password-authenticator.properties`, or run the [real tests](https://github.com/weissleb/presto/blob/salesforce-authn-clean/presto-password-authenticators/src/test/java/io/prestosql/plugin/password/salesforce/TestSalesforceBasicAuthenticator.java#L155) by setting the environment variables.

@lhofhansl

dain commented on presto-password-authenticators/src/test/java/io/prestosql/plugin/password/salesforce/TestSalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,269 @@
+    {
+        successResponse = "<?xml version=\"1.0\" encoding=\"UTF-8\"?><soapenv:Envelope xmlns:soapenv=\"http://schemas.xmlsoap.org/soap/envelope/\" xmlns=\"urn:partner.soap.sforce.com\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"><soapenv:Body><loginResponse><result><metadataServerUrl>https://example.salesforce.com/services/Soap/m/46.0/example</metadataServerUrl><passwordExpired>false< ...
+        failedResponse = "<?xml version=\"1.0\" encoding=\"UTF-8\"?><soapenv:Envelope xmlns:soapenv=\"http://schemas.xmlsoap.org/soap/envelope/\" xmlns:sf=\"urn:fault.partner.soap.sforce.com\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"><soapenv:Body><soapenv:Fault><faultcode>sf:INVALID_LOGIN</faultcode><faultstring>INVALID_LOGIN: Invalid username, password, security token; or user lo ...
+        forReal = false;
```
This technique for enabling full integration tests doesn't look like other tests in Presto I have seen.  I could be wrong, but I would look at the other tests and copy the techniques.

dain commented on presto-password-authenticators/src/test/java/io/prestosql/plugin/password/salesforce/TestSalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,269 @@
+    private String password;
+    private String successResponse;
+    private String failedResponse;
+    private SalesforceConfig config;
```
Presto runs test in parallel with multiple threads, and I don't think this is thread safe.

I suggest removing all of the fields here and up, and just using local variables.

dain commented on presto-password-authenticators/src/test/java/io/prestosql/plugin/password/salesforce/TestSalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,269 @@
+        if (org == null || org.length() == 0) {
+            fail("Must set SALESFORCE_TEST_ORG environment variable.");
+        }
+        username = System.getenv("SALESFORCE_TEST_USERNAME");
```
You can wrapper all of these with `emptyToNull` to simplify the usages

dain commented on presto-password-authenticators/src/test/java/io/prestosql/plugin/password/salesforce/TestSalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,269 @@
+        }
+        username = System.getenv("SALESFORCE_TEST_USERNAME");
+        password = System.getenv("SALESFORCE_TEST_PASSWORD");
+        if (username == null || username.length() == 0 || password == null || password.length() == 0) {
```
this form `username.length() == 0` should be replaced with `username.isEmpty()`. Your IDE should be making this suggestion for you automatically.

weissleb replied:
Didn't get any hint on it from IntelliJ, but will research perhaps why.  I've replaced all these checks with `emptyToNull` as suggested.  These checks are out of old habit.  I should form new habits.

dain commented on presto-password-authenticators/src/test/java/io/prestosql/plugin/password/salesforce/TestSalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,269 @@
+
+        testHttpClient = new TestingHttpClient((request -> mockResponse(HttpStatus.INTERNAL_SERVER_ERROR, MediaType.ANY_TEXT_TYPE, xmlResponse)));
+        SalesforceBasicAuthenticator authenticator = new SalesforceBasicAuthenticator(config, testHttpClient);
+        Principal principal = authenticator.createAuthenticatedPrincipal(username, password);
```
(nit) removed unused variable `principal`

dain commented on presto-password-authenticators/src/test/java/io/prestosql/plugin/password/salesforce/TestSalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,269 @@
+    }
+
+    @Test(expectedExceptions = AccessDeniedException.class)
+    public void createAuthenticatedPrincipal_wrongOrg()
```
(nit) I don't think we use underscores in method names in Presto

dain commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceConfig.java:
```diff
@@ -0,0 +1,115 @@
+    private static final Logger log = Logger.get(SalesforceConfig.class);
+
+    protected static final String LOGINURL = "https://login.salesforce.com/services/Soap/u/";
+    protected static final String APIVERSION = "46.0";
```
(nit) Add underscores to the names to the spell checker in the ID doesn't complain

dain commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceConfig.java:
```diff
@@ -0,0 +1,115 @@
+
+public class SalesforceConfig
+{
+    private static final Logger log = Logger.get(SalesforceConfig.class);
```
This isn't used

dain commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceConfig.java:
```diff
@@ -0,0 +1,115 @@
+    public static final int MAX_EXPIRE = 3600;
+    private int cacheSize = 4096;
+    private int cacheExpireSeconds = 120;
+    private String orgs;
```
Also rename to `allowedOrganizations`

weissleb replied:
Good idea on the rename.  

I originally had the default just implicitly be "all", but @lhofhansl strongly suggested that we require admins to set this explicitly.  I agree, and not just cause he's my boss.  I actually raised the concern about implicitly allowing all orgs, and do in fact prefer the explicit setting.

dain commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceConfig.java:
```diff
@@ -0,0 +1,115 @@
+            " </env:Body>\n" +
+            "</env:Envelope>\n";
+
+    public static final int MAX_EXPIRE = 3600;
```
This can be private

weissleb replied:
Originally didn't depend on the annotations for validation, and therefore used this throughout the package.  I've refactored to use `@MaxDuration` per your suggestion.

weissleb commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,187 @@
+{
+    private static final Logger log = Logger.get(SalesforceBasicAuthenticator.class);
+
+    private Set<String> orgs;                                   // Set of Salesforce orgs, which users must belong to in order to authN.
```
No problem.  Updating.

weissleb commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,187 @@
+{
+    private static final Logger log = Logger.get(SalesforceBasicAuthenticator.class);
+
+    private Set<String> orgs;                                   // Set of Salesforce orgs, which users must belong to in order to authN.
```
Solid idea.  Will do.

weissleb commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,187 @@
+    @Override
+    public Principal createAuthenticatedPrincipal(String username, String password)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
```
Not strictly, no I don't think so.  Original implementation did not include this, but honestly I added it after reviewing the LDAPAuthenticator.

weissleb commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,187 @@
+    // This does the work of logging into Salesforce.
+    private Principal doLogin(Credential credential)
+    {
+        log.info("Logging into Salesforce.");
```
Had it there for a couple of reasons.  First is simply to ensure the cache was working as expected, so you're right, definitely not a production need.  The second is that at Salesforce we are using multiple authentication options, so that users can login either via credentials we store in a password file, or through another custom mechanism we wrote, or via their Salesforce credentials.  This is a little bit of a crude audit mechanism, though clearly very crude.

Happy to remove it, or is it acceptable to change to `debug`?

weissleb commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,187 @@
+        StringResponseHandler.StringResponse response = httpClient.execute(request, StringResponseHandler.createStringResponseHandler());
+
+        final int statusCode = response.getStatusCode();
+        if (statusCode < 200 || statusCode >= 300) {
```
You're right.  Response should be 200.

weissleb commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceConfig.java:
```diff
@@ -0,0 +1,115 @@
+    // and responses are English, and organization id is not in Japaneses characters (this is
+    // also true of the organization id in the UI, even with other text showing in Japanese).
+
+    @NotNull(message = "Must set salesforce.org with one or more Salesforce 18 char OrgId's, or \"all\"")
```
So, OrgId is commonly used among Salesforce employees and admins.  That stated, it's not found in official documentation I don't think.  I'll change it to Organization Id, and will fix the possessive.

weissleb commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,187 @@
+                    e.getMessage()));
+        }
+
+        // Parse the response into a flattened map.
```
Nice!  Should work.

weissleb commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,187 @@
+{
+    private static final Logger log = Logger.get(SalesforceBasicAuthenticator.class);
+
+    private Set<String> orgs;                                   // Set of Salesforce orgs, which users must belong to in order to authN.
```
Will get rid of `responseMap`, per your rewrite of the document parsing.

weissleb commented:
@dain thank you for the very thorough review.  I've addressed all items and updated the PR with another commit. 

One exception is your point about the [integration tests](https://github.com/prestosql/presto/pull/4372#discussion_r456851812).  I will try to find other examples in the project and see how I can conform.

Please advise on whether you'd like me to resolve the comments.  Much appreciated!

findepi commented:
> One exception is your point about the [integration tests](https://github.com/prestosql/presto/pull/4372#discussion_r456851812). I will try to find other examples in the project and see how I can conform.

We have following tests interfacing with external services:

* S3 tests https://github.com/prestosql/presto/blob/bc9790eef2a4664e0ca339167581447b6b24a9e4/.github/workflows/ci.yml#L91-L100
* Azure tests https://github.com/prestosql/presto/pull/3044
* Google Sheets connector tests (`TestGoogleSheets`)

Of these, the last case seems to be the best example in this case.

We also have a swathe of docker-based tests for most of our integration testing.
Not sure if there is a dockerized Salesforce exposing same APIs?

weissleb commented:
> > One exception is your point about the [integration tests](https://github.com/prestosql/presto/pull/4372#discussion_r456851812). I will try to find other examples in the project and see how I can conform.
> 
> We have following tests interfacing with external services:
> 
> * S3 tests https://github.com/prestosql/presto/blob/bc9790eef2a4664e0ca339167581447b6b24a9e4/.github/workflows/ci.yml#L91-L100
> * Azure tests #3044
> * Google Sheets connector tests (`TestGoogleSheets`)
> 
> Of these, the last case seems to be the best example in this case.
> 
> We also have a swathe of docker-based tests for most of our integration testing.
> Not sure if there is a dockerized Salesforce exposing same APIs?

@findepi thank you!  I've taken a look, and see what you guys are doing.  I don't think this necessarily warrants making it's way into the CI tests like you have with the connectors, since it's simply testing a login.  I'm checking with our auth team to see if there is a non-expiring user with no permissions that we could provide similar to the Google Sheets test.

My intent was simply to allow admins to easily test and understand the plugin before deploying it, though it's pretty straightforward.  If we aren't able to find credentials we can provide into the project, and there's no appetite for adopting the style of testing I've implemented here, I am fine removing those "for real" tests.

weissleb commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,185 @@
+                " <env:Body>\n" +
+                "   <n1:login xmlns:n1=\"urn:partner.soap.sforce.com\">\n" +
+                "     <n1:username>%s</n1:username>\n" +
+                "     <n1:password>%s</n1:password>\n" +
```
@dain something as simple as this?

```
Escaper escaper = xmlContentEscaper();
        Request request = new Request.Builder()
                .setUri(URI.create(loginUrl + apiVersion))
                .setHeader("Content-Type", "text/xml;charset=UTF-8")
                .setHeader("SOAPAction", "login")
                .setMethod("POST")
                .setBodyGenerator(createStaticBodyGenerator(String.format(loginSoapMessage, escaper.escape(username), escaper.escape(password)), UTF_8))
                .build();
```

weissleb commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceBasicAuthenticator.java:
```diff
@@ -0,0 +1,185 @@
+        String returnedOrg = getElementValue(xmlResponse, "organizationId");
+        // If the only entry in the set is "all", don't bother to check, otherwise make sure the returned org is in the set.
+        // The organizationId is always in Locale.US, regardless of the user's locale and language.
+        if (!(allowedOrganizations.size() == 1 && allowedOrganizations.contains("all"))
```
Updating check as follows.

```
// If the only entry in the set is "all", don't bother to check, otherwise make sure the returned org is in the set.
        // The organizationId is always in Locale.US, regardless of the user's locale and language.
        if (!allowedOrganizations.equals(ImmutableSet.of("all"))) {
            if (!allowedOrganizations.contains(returnedOrg.toLowerCase(Locale.US))) {
                throw new AccessDeniedException(String.format(
                        "Login successful, but for wrong Salesforce org.  Got %s, but expected a different org.",
                        returnedOrg));
            }
        }
```

weissleb commented on presto-password-authenticators/src/main/java/io/prestosql/plugin/password/salesforce/SalesforceConfig.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    @Config("salesforce.cache-expire-seconds")
+    @ConfigDescription("Expire time in seconds for an entry in cache since last write.  Default is 120 seconds, max is 3600 seconds.")
```
Cool?

```
@MaxDuration(value = "1h")
    public Duration getCacheExpireDuration()
    {
        return cacheExpireDuration;
    }

    @Config("salesforce.cache-expire-duration")
    @ConfigDescription("Expire duration for an entry in cache since last write.")
    public SalesforceConfig setCacheExpireDuration(Duration cacheExpireDuration)
    {
        this.cacheExpireDuration = cacheExpireDuration;
        return this;
    }
```

weissleb commented:
> Looks good.

Awesome, thank you @dain !  What's the next step here?  Looks like all checks are passing now.  Do we still need review from @electrum ?

dain commented:
@weissleb sorry for the delay. It is merged now.  Thanks!

dain commented:
@weissleb Sorry I forgot, we should have docs for this.  Can you add docs similar to https://prestosql.io/docs/current/security/password-file.html (source https://github.com/prestosql/presto/blob/master/presto-docs/src/main/sphinx/security/password-file.rst)?

weissleb commented:
> @weissleb Sorry I forgot, we should have docs for this. Can you add docs similar to https://prestosql.io/docs/current/security/password-file.html (source https://github.com/prestosql/presto/blob/master/presto-docs/src/main/sphinx/security/password-file.rst)?

Thank you sir!  Absolutely I can add docs.  Should I just open a new PR for them, @dain ?
